tiledb/common/memory_tracker.cc

```diff
@@ -0,0 +1,319 @@
+  return ret.get();
+}
+
+std::tuple<uint64_t, std::unordered_map<MemoryType, uint64_t>>
+MemoryTracker::get_counts() {
+  std::lock_guard<std::mutex> lg(mutex_);
+
+  auto total = total_counter_.fetch_add(0, std::memory_order_relaxed);
+  std::unordered_map<MemoryType, uint64_t> by_type;
+  std::vector<MemoryType> to_del;
+  for (auto& [mem_type, resource] : resources_) {
+    by_type[mem_type] = resource->get_count();
+  }
+
+  return {total, by_type};
+}
+
+uint64_t MemoryTracker::generate_id() {
+  static std::atomic<uint64_t> curr_id{0};
+  return curr_id.fetch_add(1);
+}
+
+shared_ptr<MemoryTracker> MemoryTrackerManager::create_tracker() {
+  class MemoryTrackerCreator : public MemoryTracker {
+   public:
+    MemoryTrackerCreator()
+        : MemoryTracker() {
+    }
+  };
+
+  std::lock_guard<std::mutex> lg(mutex_);
+
+  // Delete any expired weak_ptr instances
+  size_t idx = 0;
+  while (idx < trackers_.size()) {
+    if (trackers_[idx].expired()) {
+      trackers_.erase(trackers_.begin() + idx);
+    } else {
+      idx++;
+    }
+  }
+
+  // Create a new tracker
+  auto ret = make_shared<MemoryTrackerCreator>(HERE());
+  trackers_.emplace(trackers_.begin(), ret);
+
+  return ret;
+}
+
+std::string MemoryTrackerManager::to_json() {
+  std::lock_guard<std::mutex> lg(mutex_);
+  nlohmann::json rv;
+
+  size_t idx = 0;
+  while (idx < trackers_.size()) {
+    auto ptr = trackers_[idx].lock();
+    if (!ptr) {
+      trackers_.erase(trackers_.begin() + idx);
+      continue;
+    }
+
+    nlohmann::json val;
+
+    // Set an distinguishing id
+    val["tracker_id"] = std::to_string(ptr->get_id());
+
+    // Mark the stats with the tracker type.
+    val["tracker_type"] = memory_tracker_type_to_str(ptr->get_type());
+
+    // Add memory stats
+    auto [total, by_type] = ptr->get_counts();
+    val["total_memory"] = total;
+    val["by_type"] = nlohmann::json::object();
+    for (auto& [type, count] : by_type) {
+      val["by_type"][memory_type_to_str(type)] = count;
+    }
+    rv.push_back(val);
+
+    idx++;
+  }
+
+  return rv.dump();
+}
+
+MemoryTrackerReporter::~MemoryTrackerReporter() {
+  if (!filename_.has_value()) {
+    return;
+  }
+
+  {
+    std::lock_guard<std::mutex> lg(mutex_);
+    stop_ = true;
+    cv_.notify_all();
+  }
+  try {
+    thread_.join();
+  } catch (std::exception& exc) {
+    LOG_ERROR(
+        "Error stopping MemoryTrackerReporter thread: " +
+        std::string(exc.what()));
+  }
+}
+
+void MemoryTrackerReporter::start() {
+  if (!filename_.has_value()) {
+    LOG_INFO("No filename set, not starting the MemoryTrackerReporter.");
+    return;
+  }
+
+  {
+    // Scoped so we release this before the thread starts. Probably unnecessary
+    // by better safe than sorry.
+    std::lock_guard<std::mutex> lg(mutex_);
+    if (stop_) {
+      throw std::runtime_error("MemoryTrackerReporters cannot be restarted.");
+    }
+  }
+
+  // Thread start logic mirrored from the ThreadPool.
```

Comment:
This code is now "copy pasted" 5 times... We should implement a common method for this. Note that I'm opened for a follow up story that we do next iteration.